src/controls/LevelingSensor/QskLevelingSensorSkinlet.cpp

```diff
@@ -0,0 +1,435 @@
+            , sX( r1 / sensor->angle().x() )
+            , sY( r1 / sensor->angle().y() )
+            , sZ( r1 / sensor->angle().z() )
+        {
+        }
+
+        Q_REQUIRED_RESULT QVector3D scale() const noexcept 
+        {
+            return {(float)sX, (float)sY, (float)sZ};
+        }
+    };
+
+    template<>
+    struct State< QskAspect::Subcontrol > : State< QskLevelingSensor >
+    {
+        qreal rX = 0.0;
+        qreal rY = 0.0;
+        qreal rZ = 0.0;
+        qreal tX = 0.0;
+        qreal tY = 0.0;
+        qreal tZ = 0.0;
+
+        State( const QskLevelingSensor* const sensor, QskAspect::Subcontrol subcontrol )
+            : State< QskLevelingSensor >( sensor )
+            , rX( sensor->subControlRotation( subcontrol ).x() )
+            , rY( sensor->subControlRotation( subcontrol ).y() )
+            , rZ( sensor->subControlRotation( subcontrol ).z() )
+            , tX( rY * sX )
+            , tY( rX * sY )
+            , tZ( 0.0 )
+        {
+        }
+
+        Q_REQUIRED_RESULT QVector3D translation() const noexcept 
+        {
+            return {(float)tX, (float)tY, (float)tZ};
+        }
+
+        Q_REQUIRED_RESULT QMatrix4x4 matrix() const noexcept
+        {
+            return matrix_deg( 0, 0, rZ, cX, cY );
+        }
+    };
+
+}
+
+using Q = QskLevelingSensor;
+using R = QskLevelingSensorSkinlet::NodeRole;
+
+using namespace QskSGNode;
+
+template< typename Root, typename... Children >
+inline Q_REQUIRED_RESULT Root* ensureNodes( QSGNode* root = nullptr )
+{
+    return ensureNodes<AppendMode::Recursive, Root, Children...>(root);
+}
+
+float QskLevelingSensorSkinlet::outerRadius( const QskSkinnable* const skinnable )
+{
+    const auto* const sensor = static_cast< const Q* >( skinnable );
+    const auto contentsRect = sensor->contentsRect();
+    return 0.5f * qMin( contentsRect.width(), contentsRect.height() );
+}
+
+float QskLevelingSensorSkinlet::innerRadius( const QskSkinnable* const skinnable )
+{
+    const auto scale = skinnable->strutSizeHint( Q::Horizon );
+    return outerRadius( skinnable ) * scale.width();
+}
+
+QPointF QskLevelingSensorSkinlet::center( const QskSkinnable* const skinnable )
+{
+    const auto* const sensor = static_cast< const Q* >( skinnable );
+    return sensor->contentsRect().center();
+}
+
+QskLevelingSensorSkinlet::QskLevelingSensorSkinlet( QskSkin* skin )
+    : Inherited( skin )
+{
+    setNodeRoles( {
+        OuterDisk,
+        Horizon,
+        HorizonClip,
+        TickmarksX,
+        TickmarksXLabels,
+        TickmarksY,
+        TickmarksYLabels,
+        TickmarksZ,
+        TickmarksZLabels,
+    } );
+}
+
+template< QskLevelingSensorSkinlet::NodeRole >
+Q_REQUIRED_RESULT QRectF QskLevelingSensorSkinlet::subControlRect(
+    const QskLevelingSensor* const sensor, const QRectF& contentsRect ) const = delete;
+
+template<>
+Q_REQUIRED_RESULT QRectF QskLevelingSensorSkinlet::subControlRect< R::OuterDisk >(
```

Comment:
I wonder if this template specialization is needed here

Reply:
not needed, but subjectively nice to look at ;)